### DISCUSSION TITLE
About use the backup method to call the original method

### DISCUSSION THREAD

**Discussion by buffcow** [LSPosed/LSPosed#1028] [Q&A] [answered]
Hello every devs, I want to ask a question about development here.
Regarding the [LspHooker class](https://github.com/LSPosed/LSPosed/blob/13a8dd768599f7656cdf6d4d114b93f19b48c493/core/src/main/java/de/robv/android/xposed/LspHooker.java) of [handleHookedMethod](https://github.com/LSPosed/LSPosed/blob/13a8dd768599f7656cdf6d4d114b93f19b48c493/core/src/main/java/de/robv/android/xposed/LspHooker.java#L45) method where the parameter is args, If the method is non-static the args will have a "thisObject". LspHooker class has a backup field. hook method if non-static the first parameter of the backup method is thisObject, but “param.args” is processed:
`if (Modifier.isStatic(method.getModifiers())) {`
`           ....`
`        } else {`
`            param.thisObject = args[0];`
`            param.args = new Object[args.length - 1];`
`            System.arraycopy(args, 1, param.args, 0, args.length - 1);`
`        }`

But why in [here](https://github.com/LSPosed/LSPosed/blob/13a8dd768599f7656cdf6d4d114b93f19b48c493/core/src/main/java/de/robv/android/xposed/LspHooker.java#L93) is invoke by "param.args", not the "args" parameter of the handleHookedMethod function? Isn't “param.args" already subtracted a thisObject?

I invoke like you too but the carsh log prompt is one less parameter:
`java.lang.IllegalArgumentException: Wrong number of arguments; expected 4, got 3`
I invoke backup method by "args"  is ok:
`backup.invoke(param.thisObject, args)`
Why?
@yujincheng08

**Comment by yujincheng08** [comment, marked as answer]
If the original method is static, then backup should be invoked by `invoke(null, args)`.
If the original method is non-static, then the backup should be invoked by `invoke(thisObject, args)`.

For Xposed interface, `thisObject` is always null (unless some modules set it) for static methods. So both invokes can be merged to `invoke(thisObject, args)`

**Reply by buffcow** [reply]
yeah  i know.  I understand the thisObject
But I was wondering why args can use "param.args" instead of the "args" inside the `invokeOriginalMethod (Object thisObject, Object [] args)` 
The first index of the param.args is no longer a thisObject!  The first parameter of the backup method is thisObject!

**Reply by yujincheng08** [reply]
I don't understand your question. both param.args and args from `invokeOriginalMethod` have no thisObject.

**Reply by yujincheng08** [reply]
> But why in here is invoke by "param.args", not the "args" parameter of the handleHookedMethod function? Isn't “param.args" already subtracted a thisObject?

Why not using `param.args` who does not contain thisObject? From `invoke`'s document, the first parameters should be the `this` and the rest should be the parameters without `this`.

**Reply by buffcow** [reply]
ahhh    It's all right,  I take an example
```
class Test{
  public void test(String str,int num){...}
}
```
For example I want the hook the test method then the dexBuilder generated method is
`backup(Object thiz, String str, int num){...}`

if i invoke as this >> `backup.invoke(param.thisObject,param.args)`
Then the first index of  **`param.args`** is `String` , is not `thiz`.
because [here](https://github.com/LSPosed/LSPosed/blob/13a8dd768599f7656cdf6d4d114b93f19b48c493/core/src/main/java/de/robv/android/xposed/LspHooker.java#L56)   
param.args[0] is `String` , [1] is `int`

**Reply by yujincheng08** [reply]
That's all right. I don't see any problem. The invoke is great.

**Reply by buffcow** [reply]
yeah   But I call like that will be missing parameters  of `thiz`  ，i don't know where the reason

**Reply by buffcow** [reply]
Anyway, I do the research myself
### thank you very much！

**Reply by yujincheng08** [reply]
have u used it as a backup of YAHFA?

**Reply by buffcow** [reply]
y .   the code is the same as you wrote.
for non-static methods, the first argument of the yahfa's backup method is always `thiz`..


About this code:

> `param.thisObject = args[0];`
> `param.args = new Object[args.length - 1];`
> `System.arraycopy(args, 1, param.args, 0, args.length - 1);`

as you can see from the above code: `param.args` has only the actual original parameter array, which does not contain the `thiz` parameter.

If I invoke the backup method with `param.args`, the first parameter of `thiz` is missing.

and then will cause >> 
`java.lang.IllegalArgumentException: Wrong number of arguments; expected n+1, got n`.

I don't know why it is normal in LSP, it makes me confused(°ー°〃)

**Reply by yujincheng08** [reply]
Note that backup method will become non-statuc after it is used to backup a non-static method. Its name, declaring class will be totally changed to be the same as the original method rather how its generated. So at that time, invoking it is like invoking a non-static method. In your case I think its because the backup method is NOT used as a backup and it is still a static method.

**Reply by buffcow** [reply]
Okay, even if it is a non-static backup method, the parameter always has a thiz, isn't it?
yahfa example：
<img width="545" alt="backup" src="https://user-images.githubusercontent.com/42611305/131456865-676eb1df-5888-4593-b9aa-c4ad1eaf249e.png">
Let's take a look at the[`mActualParameterTypes`](https://github.com/LSPosed/LSPosed/blob/13a8dd768599f7656cdf6d4d114b93f19b48c493/core/src/main/java/org/lsposed/lspd/yahfa/dexmaker/HookerDexMaker.java#L110)：
<img width="526" alt="mActualParameterTypes" src="https://user-images.githubusercontent.com/42611305/131457945-2746aac1-baa1-45cf-be13-881b0534e4ff.png">
**then** 
<img width="529" alt="getParameterTypes" src="https://user-images.githubusercontent.com/42611305/131457993-2c64ca8d-dc3a-4704-be10-865d3e9aa4d8.png">

So if you hook a non-static method, the backup method always has a `thiz` parameter, right?

Let us continue to look at this code in the [`LspHooker`](https://github.com/LSPosed/LSPosed/blob/13a8dd768599f7656cdf6d4d114b93f19b48c493/core/src/main/java/de/robv/android/xposed/LspHooker.java#L55) class:
<img width="350" alt="param" src="https://user-images.githubusercontent.com/42611305/131459032-ba0107a8-0cb4-4f04-abf2-5d2b4a2a0efd.png">
**then in [here](https://github.com/LSPosed/LSPosed/blob/13a8dd768599f7656cdf6d4d114b93f19b48c493/core/src/main/java/de/robv/android/xposed/LspHooker.java#L93)**
<img width="370" alt="invoke_backup" src="https://user-images.githubusercontent.com/42611305/131459276-d3c64653-dd2b-4b8e-bb4d-41c0ad7de31e.png">

`param.args` has no `thiz` object. 
Why did you use it as  parameters to invoke the backup method without `thiz` and succeed? And when I invoke, I get an error saying that I am missing a `thiz` parameter?＞︿＜

**Reply by yujincheng08** [reply]
it's because you don't change `backup` to non-static.

**Reply by buffcow** [reply]
Oh? Isn't `backup `static by default? Does a non-static `backup` throw an exception? [here](https://github.com/LSPosed/LSPosed/blob/4cb731c1fb8cff57caa17c7e2376619b0a87bfb8/core/src/main/java/org/lsposed/lspd/core/yahfa/HookMain.java#L49)
<img width="480" alt="ckbak" src="https://user-images.githubusercontent.com/42611305/131497800-5a590078-c1ca-4df2-ac6e-f756b7682ec0.png">

**Reply by yujincheng08** [reply]
Recall what I said before.

> Note that the backup method will become non-static after it is used to backup a non-static method.

**Reply by buffcow** [reply]
ahh I am really confused.😥
Can you give me an example.?
For example, if I want to hook `public void test(String str)`, what will be the `backup` method generated by dexBuilder?

**Reply by buffcow** [reply]
> Note that the backup method will become non-static after it is used to backup a non-static method.

我不明白静态和非静态的`backup` 和调用参数有啥关系
第一个参数不都是`thiz`吗😶
大佬  我真的很懵<(＿　＿)>

**Reply by yujincheng08** [reply]
For example, you are hooking:
```java
class Test {
  public void test(String str) {}
  public void testStatic(String str){}
}
```

If these two methods are not hooked, we reflectively invoke them like:
```java
testMethod.invoke(testObject, someString);
testStaticMethod.invoke(null, someString);
```

The dex builder will generate the backup methods like:
```java
static void backupTest(Test testObject, String str);
static void backupTestStatic(String str);
```

If these methods are not yet used to backup, reflectively you should invoke like:
```java
backupTestMethod.invoke(null, testObject, someString);
backupTestStaticMethod.invoke(null, someString);
```

After they are used to backup methods, you should invoke them like my second code snippets as if they are the reflection of the hooked methods:
```java
backupTestMethod.invoke(testObject, someString); // testMethod.invoke(testObject, someString);
backupTestStaticMethod.invoke(null, someString); // testStaticMethod.invoke(null, someString);
```

As you can see, before and after the `backupTestMethod` is used to backup, the invocation is different.

So if you use the `backupTestMethod` BEFORE it's used to backup by `backupTestMethod.invoke(testObject, someString);`, you will encounter `IllegalArgumentException`.

**Reply by buffcow** [reply]
> So if you use the backupTestMethod BEFORE it's used to backup by backupTestMethod.invoke(testObject, someString);, you will encounter IllegalArgumentException.

oh!🧐   But where is the code of ` used as a backup`?

**Reply by buffcow** [reply]
> As you can see, before and after the backupTestMethod is used to backup, the invocation is different.

Is the `used to backup` yahfa's [`backupAndHook`](https://github.com/PAGalaxyLab/YAHFA/blob/fe3971295f149068664a28be6589f9d28bcedc86/library/src/main/java/lab/galaxy/yahfa/HookMain.java#L113)?

**Reply by yujincheng08** [reply]
Yup. But YAHFA is different from ours.

**Reply by buffcow** [reply]
ahhh.. I'm using yahfa's original code, it's not yours 
Now I finally understand, thank you very much!😃
